Standards/scs-0214-v2-k8s-node-distribution.md

```diff
@@ -0,0 +1,129 @@
+
+## Motivation
+
+In normal day-to-day operation, it is not unusual for some operational failures, either
+due to wear and tear of hardware, software misconfigurations, external problems or
+user errors. Whichever was the source of such an outage, it always means down-time for
+operations and users and possible even data loss.
+Therefore, a Kubernetes cluster in a productive environment should be distributed over
+multiple "failure zones" in order to provide fault-tolerance and high availability.
+This is especially important for the control plane of the cluster, since it contains the
+state of the whole cluster. A failure of this component could mean an unrecoverable failure
+of the whole cluster.
+
+## Design Considerations
+
+Most design considerations of this standard follow the previously written Decision Record
+[Kubernetes Nodes Anti Affinity][scs-0213-v1] as well as the Kubernetes documents about
+[High Availability][k8s-ha] and [Best practices for large clusters][k8s-large-clusters].
+
+SCS wishes to prefer distributed, highly-available systems due to their obvious advantages
+like fault-tolerance and data redundancy. But it also understands the costs and overhead
+for the providers associated with this effort, since the infrastructure needs to have
+hardware which will just be used to provide fail-over safety or duplication.
+
+The document [Best practices for large clusters][k8s-large-clusters] describes the concept of a failure zone.
+This term isn't defined any further, but can in this context be described as a number of
+physical (computing) machines in such a vicinity to each other (either through physical
+or logical interconnection in some way), that specific problems inside this zone would put
+all these machines at risk of failure/shutdown. It is therefore necessary for important
+data or services to not be present just on one failure zone.
+How such a failure zone should be defined is dependent on the risk model of the service/data
+and its owner as well as the capabilities of the provider. Zones could be set from things
+like single machines or racks up to whole datacenters or even regions, which could be
+coupled by things like electrical grids. They're therefore purely logical entities, which
+shouldn't be defined further in this document.
+
+## Decision
+
+This standard formulates the requirement for the distribution of Kubernetes nodes in order
+to provide a fault-tolerant and available Kubernetes cluster infrastructure.
+
+The control plane nodes MUST be distributed over multiple physical machines.
+Kubernetes provides [best-practices][k8s-zones] on this topic, which are also RECOMMENDED by SCS.
+
+At least one control plane instance MUST be run in each "failure zone" used for the cluster,
+more instances per "failure zone" are possible to provide fault-tolerance inside a zone.
```

Comment:
I experienced an interesting scenario, where the node distribution test will succeed.
**SCS Compliance results** in https://github.com/SovereignCloudStack/k8s-cluster-api-provider/pull/742#issuecomment-2172470815 show `INFO: The nodes are distributed across 2 host-ids.`
I am wondering now if 3 control plane nodes are distributed over 2 hosts, it means that 2 control plane nodes are located on the same physical host. What will happen with the k8s cluster if this host goes down? AFAIK etcd has in this case only one member failure tolerance, it will stop functioning!

Reply:
These lines in the standard and corresponding test don't require hard anti-affinity.

Reply:
I understand the concern. We would probably need to adapt the lines above the ones you mentioned from
```
The control plane nodes MUST be distributed over multiple physical machines.
Kubernetes provides [best-practices][k8s-zones] on this topic, which are also RECOMMENDED by SCS.
```
to
```
The control plane nodes MUST be distributed over multiple physical machines; a control plane MUST contain at least three nodes and be distributed over three machines.
Kubernetes provides [best-practices][k8s-zones] on this topic, which are also RECOMMENDED by SCS.
```

Would that be alright?

Reply:
`at least three nodes` can this be considered as node distribution standard or is it more suited for some k8s HA standard?

Reply:
Also, can we somehow incorporate the fact that etcd can be external? Then we are probably good with the current formulation, this comment is valid only if stacked etcd for k8s cluster is used.

Reply:
Probably more relevant for a `HA` standard, but TBF that was always the problem I had with this standard in general, since `node distribution` or better `fault tolerance` also kind of (at least partly) includes `(high)  availability`.
Personally, after working on this standard this long I would probably change it up altogether to create a `(high) availability` that is based on this document and make this `distribution` standard optional, since not every cluster needs to be distributed (IMO).


Coming back to the case presented by you above:
> I am wondering now if 3 control plane nodes are distributed over 2 hosts, it means that 2 control plane nodes
> are located on the same physical host. What will happen with the k8s cluster if this host goes down?
> AFAIK etcd has in this case only one member failure tolerance, it will stop functioning!
I understand this scenario (and as far as I know you're right about the etcd functionality). The question would be how to fix this scenario in the standard. There  shouldn't be a number of physical nodes smaller than three, if etcd in use has 3 members.
So I guess this means this is still a distribution issue for us and not a `HA` issue?

Reply:
> Also, can we somehow incorporate the fact that etcd can be external? Then we are probably good with the current formulation, this comment is valid only if stacked etcd for k8s cluster is used.

Now I am [reading](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/#external-etcd-topology) that also for external etcd you need 3+3 nodes. Do you know why it is so?

> I understand this scenario (and as far as I know you're right about the etcd functionality). The question would be how to fix this scenario in the standard. There  shouldn't be a number of physical nodes smaller than three, if etcd in use has 3 members.
So I guess this means this is still a distribution issue for us and not a `HA` issue?

https://kubernetes.io/docs/setup/best-practices/multiple-zones/#control-plane-behavior mentions:
```
If availability is an important concern, select at least three failure zones and replicate each individual control plane component (API server, scheduler, etcd, cluster controller manager) across at least three failure zones.
```
So maybe we can keep it and put it only later to the HA standard?

Reply:
Interesting, I already read about this some time ago.
I guess the idea behind the `external` etcd is separation of concerns in your infrastructure and better control (and possible security) for a cluster. So it makes sense to separate the underlying etcd nodes and the kubernetes control plane. As the linked article mentions
```
This topology decouples the control plane and etcd member. It therefore provides an HA setup where losing a
control plane instance or an etcd member has less impact and does not affect the cluster redundancy
as much as the stacked HA topology.
```
Now I think that use-case wouldn't be implemented that often, but its an interesting thing to consider for a setup with required redundancy.

Personally, I would put it into a future `High availability` standard.

I will mention all this in #639 and we can keep on discussing everything there.